Add HomeKit Television functionality

adrum opened the pull request:
## Description:

CC: @cdce8p 

This adds the ability to expose `media_player` entities as Television types in HomeKit. This is an opt-in feature, requiring users with devices iOS 12.2 or later to change the `media_player` entity's `device_class: tv`.

I plan to add the ability to link a `remote` later on for additional functionality. 

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#9186

## Example entry for `configuration.yaml` (if applicable):
```yaml
    media_player.bedroom:
      device_class: tv
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-manifest]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json
[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json#L5
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23
[manifest-docs]: https://developers.home-assistant.io/docs/en/development_checklist.html#_the-manifest-file_

adrum committed:
added tv from media player

adrum committed:
Merge branch 'dev' into add-homekit-television

adrum committed:
added play_pause for homekit tv

adrum committed:
added warning and checks on None for sources

adrum committed:
removed initial value for on_off

adrum committed:
added type validation

adrum committed:
added television test

adrum committed:
send specific command if known

adrum committed:
do not save play state to remotekey char

adrum committed:
fixed values

adrum committed:
fixed volume step to be testable

adrum committed:
bump HAP-python

adrum committed:
Merge branch 'dev' into add-homekit-television

# Conflicts:
#	homeassistant/components/homekit/util.py
#	tests/components/homekit/test_util.py

adrum committed:
Merge branch 'dev' into add-homekit-television

adrum committed:
remove unused service

adrum committed:
remove unused imports

adrum committed:
lint issues

adrum committed:
more lint issues

adrum committed:
updated missing dep location

adrum committed:
fixed a test

cdce8p commented:
@adrum I'll probably start the review this weekend. In the meantime, could you extract the HAP-python update to a separate PR (and include the reordering of `CHAR_OUTLET_IN_USE`)? I would like this to get done first, so it's out of the way.

As for the general direction, I'm still debating if we should just make this a breaking change and support everything we can through the television type. The switches have always only been a placeholder until something better would came along.

adrum commented:
@cdce8p I've updated the version number here: #23031 and updated the ordering of constants.

As far as the direction is concerned, I'm cool with dropping support for the switches. I just thought it would increase the odds of the PR getting accepted if there wasn't a breaking change. If you think the breaking change is preferred, I can get the code ready so you can review it this weekend.

cdce8p commented:
@adrum I've merged the other PR, so that you could rebase this branch.

Regarding the direction, I might be able to take a closer look later today. I'll get back to you.

adrum committed:
Merge branch 'dev' into add-homekit-television

# Conflicts:
#	homeassistant/components/homekit/__init__.py

adrum committed:
Merge branch 'dev' into add-homekit-television

adrum committed:
fixed lint & tests

adrum commented:
@cdce8p Let me know when you've have had the chance to review this. I'm leaning towards the breaking change of removing the switches and only allowing the accessory television type.

cdce8p commented:
@adrum Sorry I didn't get back to you earlier. Although I didn't had the time to look at the PR yet, I'm also leaning towards the breaking change. This seems to be the better option here. I really hope to get to it soon, just pretty busy with university at the moment.

adrum committed:
Merge branch 'dev' into add-homekit-television-breaking-change

adrum commented:
@cdce8p I have the breaking change version ready, just in case we decide to move in that direction.

adrum committed:
Merge branch 'dev' into add-homekit-television

cdce8p commented:
@adrum I'm really really sorry, I didn't respond earlier. I finally had a chance to look at your code and it's obvious that you've spend quite some time working on it already.

As for the general direction:
I do believe we should go with the `breaking change`, or at least some version of it. The code gets overly complicated and quite difficult for me to understand if both (old and new) are intertwined. However I don't know how many users are still using iOS versions prior to 12.2 (which doesn't support the TV accessory) and, at least at the moment, I don't want to break their setup. What do you think of just having two different accessory types?

I would still plan to remove the old one one day, but for now we wouldn't have to touch it and make the new accessory type `opt in`.

Furthermore, what do you think of using the `device_class` as primary indicator? I recently saw that there had been an architecture issue opened about adding those to the `media_player` component https://github.com/home-assistant/architecture/issues/209. Since the best fit would be the `TV` class, we would exclude some entities, but I think that might for the best, at least initially.

I plan on spending more time on Home Assistant next week, so that this feature is ready for the next major release.

adrum committed:
fix typo

adrum committed:
moved telvisionmediaplayer to own accessory type

adrum committed:
map televisionmediaplayer

adrum committed:
Merge branch 'dev' into add-homekit-television

adrum committed:
removed conf_type for media players

adrum committed:
fixed tests

adrum committed:
fixed lint

adrum commented:
@cdce8p I think adding the `tv` device_class to `media_player` was a great move. I actually looked for this initially, but it wasn't there so I opted for the `entity_config` route. 

I just pushed up some changes that factors in this `device_class` and splits the tv out into its own accessory class. Hopefully, this should make it a bit easier to review.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,211 @@ def update_state(self, new_state):
+        self.chars = {FEATURE_ON_OFF: None, FEATURE_PLAY_PAUSE: None,
+                      FEATURE_TOGGLE_MUTE: None, FEATURE_SELECT_SOURCE: None,
+                      FEATURE_VOLUME_STEP: None}
+        feature_list = self.config[CONF_FEATURE_LIST]
```
If we don't require the `feature_list` as per my first comment, this block (or most of it) is probably unnecessary.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,211 @@ def update_state(self, new_state):
+
+        self.sources = []
+
+        self.category = CATEGORY_TELEVISION
```
Pass the `category` to the `super` constructor instead.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,211 @@ def update_state(self, new_state):
+        self.chars[FEATURE_ON_OFF] = television.configure_char(
+            CHAR_ACTIVE, setter_callback=self.set_on_off)
+
+        if FEATURE_PLAY_PAUSE in feature_list:
```
The feature validation should happen here now (instead of inside `utils` for the `feature_list`). You might want to take a look at `type_thermostats.py` as an example.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,211 @@ def update_state(self, new_state):
+            CHAR_ACTIVE, setter_callback=self.set_on_off)
+
+        if FEATURE_PLAY_PAUSE in feature_list:
+            self.chars[FEATURE_PLAY_PAUSE] \
```
The `dict` for `chars` was only necessary, since we basically had 4 different accessories that all mapped the same action to other chars. It would be better to name chars individually from now on, e.g.: `self.char_remote_key`.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,211 @@ def update_state(self, new_state):
+            television.add_linked_service(television_speaker)
+
+            name = '{} {}'.format(self.display_name, 'Volume')
+            television_speaker.configure_char(CHAR_NAME, value=name)
```
`name` is only used here, isn't it? Just insert it directly. No need to create a new variable.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,211 @@ def update_state(self, new_state):
+                        SERV_INPUT_SOURCE, [
+                            CHAR_IDENTIFIER,
+                            CHAR_NAME
+                        ])
```
Just one line for the four above if that fits: `SERV_INPUT_SOURCE, [CHAR_.., CHAR_NAME])`

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,211 @@ def update_state(self, new_state):
+                    _LOGGER.debug('%s: Added source %s.', self.entity_id,
+                                  source)
+            else:
+                _LOGGER.warning('%s: Source list empty when setting up sources. Please check \
```
Instead of printing a warning, we should check if the `media_player` supports different sources.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,211 @@ def update_state(self, new_state):
+                to make sure your media_player is turned on.',
+                                self.entity_id)
+
+        self.set_primary_service(television)
```
Move this line to where you first configure the `television` service above.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,211 @@ def update_state(self, new_state):
+        self.sources = []
+
+        self.category = CATEGORY_TELEVISION
+        television = self.add_preload_service(SERV_TELEVISION,
```
The service should be called something like: `serv_tv`

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,211 @@ def update_state(self, new_state):
+
+        if FEATURE_TOGGLE_MUTE in feature_list:
+
+            television_speaker = self.add_preload_service(
```
`serv_tv_speaker`?

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,211 @@ def update_state(self, new_state):
+                    CHAR_VOLUME_CONTROL_TYPE,
+                    CHAR_VOLUME_SELECTOR,
+                    CHAR_VOLUME
+                ])
```
Why use separate lines for each service?

cdce8p reviewed (changes requested):
I left some first comments.
With the change to the `device_class`, it should be no problem for us to check if the entity actually supports the features we are trying to add, thus making the `feature_list` unnecessary for `tv` entities.
However, that also means the code would have a change quite a bit. Let me know if you have questions.

FYI: I didn't looked at the `setter_callbacks` or the `update` method yet. It might make sense to hold of on updating the tests for a moment, at least until the basic structure is set.

adrum committed:
changed device class to const

adrum committed:
dont add already required characteristic

adrum committed:
remove unused feature list

adrum committed:
more fixes

adrum committed:
more simplification

adrum committed:
Merge branch 'dev' into add-homekit-television

adrum committed:
fixed device class import

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,169 @@ def update_state(self, new_state):
+            ATTR_INPUT_SOURCE)
+        if self.sources:
+            for index, source in enumerate(self.sources):
+                input_service = self.add_preload_service(SERV_INPUT_SOURCE, [
```
`serv_input`?

cdce8p commented on homeassistant/components/homekit/util.py:
```diff
@@ -36,7 +37,8 @@
         cv.string, vol.In((FEATURE_ON_OFF, FEATURE_PLAY_PAUSE,
-                           FEATURE_PLAY_STOP, FEATURE_TOGGLE_MUTE))),
+                           FEATURE_PLAY_STOP, FEATURE_TOGGLE_MUTE,
+                           FEATURE_SELECT_SOURCE, FEATURE_VOLUME_STEP))),
```
This change can probably be reverted. Same below inside `validate_media_player_features`.

cdce8p commented on homeassistant/const.py:
```diff
@@ -185,6 +185,7 @@
 DEVICE_CLASS_TIMESTAMP = 'timestamp'
 DEVICE_CLASS_PRESSURE = 'pressure'
 DEVICE_CLASS_POWER = 'power'
+DEVICE_CLASS_TV = 'tv'
```
Please move this one the `homekit/const.py` for now.

cdce8p commented on homeassistant/const.py:
```diff
@@ -318,6 +319,9 @@
 ATTR_TEMPERATURE = 'temperature'
 
+# Device classes
+DEVICE_CLASS_BATTERY
```
??

cdce8p reviewed (changes requested):
Some additional comments

awarecan commented:
FYI, `DEVICE_CLASS_TV` has been added to `media_play` in #23236

adrum committed:
renamed input service

adrum commented:
> FYI, `DEVICE_CLASS_TV` has been added to `media_player` in #23236

@cdce8p  Is there a reason this wasn't added to `media_player.const`? I know you weren't a part of that PR, but it seems odd doesn't it?

adrum committed:
dont validate TelevisionMediaPlayer accessory via config features

adrum committed:
updated char structure

adrum committed:
fixed tests

adrum committed:
lint fixes

adrum commented:
@cdce8p I took an attempt at adding characteristics based on the media players supported features. Let me know if this is or isn't what you had in mind and what you would prefer instead.

adrum committed:
condensed lines

adrum committed:
more lint and test fixes

adrum committed:
fixed tests

cdce8p commented:
> > FYI, `DEVICE_CLASS_TV` has been added to `media_player` in #23236
> 
> @cdce8p Is there a reason this wasn't added to `media_player.const`? I know you weren't a part of that PR, but it seems odd doesn't it?

Yeah it does. I guess, you could open a new PR to move them to `const` if you like. Otherwise just import from `media_player` for now.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,182 @@ def update_state(self, new_state):
+        """Send volume step value if call came from HomeKit."""
+        _LOGGER.debug('%s: Step volume by %s',
+                      self.entity_id, value)
+
```
Do we need a flag for that?

adrum replied:
I believe this is the same as play/pause. I did add a flag for tracking volume level. I still need to be able to differentiate between media players with and without volume level.

cdce8p replied:
You were right here as well. I didn't know that there is not UI inside iOS to display the current volume. The flag isn't necessary then.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,182 @@ def update_state(self, new_state):
+        """Update Television state after state changed."""
+        current_state = new_state.state
+
+        if self.char_active:
```
Since `self.char_active` is always added, this might be redundant.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,182 @@ def update_state(self, new_state):
+        current_state = new_state.state
+
+        if self.char_active:
+            hk_state = current_state not in (STATE_OFF, STATE_UNKNOWN, 'None')
```
Does `'None'` ever happen? Isn't that covered with `STATE_UNKNOWN`?

adrum replied:
I have removed the `'None'` state. This should be covered by `STATE_UNKNOWN`

adrum replied:
I did see this was also used in the normal `MediaPlayer` accessory.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,182 @@ def update_state(self, new_state):
+                self.char_active.set_value(hk_state)
+            self._flag_active = False
+
+        if self.char_toggle_mute:
```
This could throw an error if the `TV media_player` doesn't support volume controls. Instead check if the `char` has been added to the char list. Again `type_lights.py` is a good example for that.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,182 @@ def update_state(self, new_state):
+                self.char_toggle_mute.set_value(current_state)
+            self._flag_toggle_mute = False
+
+        if self.char_active_identifier:
```
Same applies here

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,182 @@ def update_state(self, new_state):
+            if self.sources and not self._flag_active_identifier:
+                _LOGGER.debug(
+                    '%s: Set current input to %s',
+                    self.entity_id, source_name)
```
This fits on the line above, barely :)

cdce8p commented on tests/components/homekit/test_util.py:
```diff
@@ -31,10 +31,10 @@ def test_validate_entity_config():
-                    {CONF_FEATURE: FEATURE_ON_OFF},
-                    {CONF_FEATURE: FEATURE_ON_OFF}]}},
+                   {CONF_FEATURE: FEATURE_ON_OFF},
+                   {CONF_FEATURE: FEATURE_ON_OFF}]}},
```
Since those white spaces didn't throw any errors originally, I would just revert the changes here.

cdce8p reviewed (changes requested):
Your last changes were a huge step in the right direction. I think this is coming together quite nicely.
I did a more detailed review this time. Let me know if you have any questions.

Just a few summarizing notes:
- I referenced `type_lights.py` one or two times as an example
- It seems as some sections of the `update_state` method have been missing? (e.g. volume control -> increase/decrease)

adrum committed:
Merge branch 'dev' into add-homekit-television

adrum commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,182 @@ def update_state(self, new_state):
+        self._flag_toggle_mute = False
+
+        # Add additional characteristics if volume or input selection supported
+        self.chars = []
```
So the reason why I didn't do this is because `ActiveIdentifier` is a required characteristic by the Television accessory. I believe it will cause errors to add this to the array passed to `add_preload_service `. Somehow, we need to be able to track the available features (select a source) for the entity in HA, but not conflict with the required characteristics of the Television accessory. I do not see any negative side effect of adding these services to the accessory, even they go unused. The `RemoteKey` char should definitely be added since it's optional.

adrum replied:
Never mind :) saw the comment below.

adrum committed:
flags

adrum committed:
one line for name

adrum committed:
serv_speaker

adrum committed:
char_mute

adrum committed:
single line for inputs

adrum committed:
revert whitespace in test util

adrum committed:
renamed chars array

adrum committed:
moved active identifier to flag

adrum committed:
removed possible error power state

adrum committed:
cleaned up sources

adrum committed:
renamed char input source

adrum committed:
moved add linked input service

adrum committed:
fixed a possible error when checking for mute status

adrum committed:
added volume char configured

adrum committed:
fixed some flags

adrum committed:
TODO: added tests for volume char

adrum commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +170,182 @@ def update_state(self, new_state):
+        params = {ATTR_ENTITY_ID: self.entity_id}
+        self.call_service(DOMAIN, service, params)
+
+    def set_play_pause(self, value):
```
To my knowledge, this looked like command sent by the `RemoteKey` char. I didn't see a way to set a state of `Playing` on any of the services or chars associated with the Television accessory.

cdce8p replied:
You're right. To me it looked like a regular callback method, initially. Since this method is only called from `set_remote_key`, wouldn't it be better to just integrate it there?

adrum committed:
removed none from state

adrum committed:
added flags for volume level

adrum committed:
fixed tests

adrum committed:
added volume level tests

adrum commented:
> > > FYI, `DEVICE_CLASS_TV` has been added to `media_player` in #23236
> > 
> > 
> > @cdce8p Is there a reason this wasn't added to `media_player.const`? I know you weren't a part of that PR, but it seems odd doesn't it?
> 
> Yeah it does. I guess, you could open a new PR to move them to `const` if you like. Otherwise just import from `media_player` for now.

It looks like the rest of the constants through HA with the prefix of `DEVICE_CLASS_` actually do not live in the `.const` for each component. So I guess this is consistent and will leave it alone.

adrum committed:
renamed setter for mute char

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,206 @@ def update_state(self, new_state):
+            self.char_active.set_value(hk_state)
+        self._flag[CHAR_ACTIVE] = False
+
+        if CHAR_VOLUME_SELECTOR in self.chars_speaker:
```
Maybe add a comment above that this part sets `mute`?

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,206 @@ def update_state(self, new_state):
+                self.char_mute.set_value(current_mute_state)
+            self._flag[FEATURE_TOGGLE_MUTE] = False
+
+        if self.support_volume_level:
```
As indicated earlier, since doesn't have a UI representation in iOS, I think it can be removed.

cdce8p reviewed (changes requested):
Some more comments

adrum committed:
moved sources

adrum committed:
added comments for updating states from HA

adrum committed:
call play pause from set_remote_key

adrum committed:
dont read volume level

adrum committed:
use in chars to check volume

adrum committed:
removed setting volume from homekit and parse state

adrum committed:
lint fixes

adrum committed:
fixed syntax

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -289,15 +282,6 @@ def set_mute(self, value):
                   ATTR_MEDIA_VOLUME_MUTED: value}
         self.call_service(DOMAIN, SERVICE_VOLUME_MUTE, params)
 
-    def set_volume(self, value):
```
You still need the callback method (`set_volume`). Only the `update_state` section for volume adjustments could be deleted.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,184 @@ def update_state(self, new_state):
+        self._flag = {CHAR_ACTIVE: False, CHAR_ACTIVE_IDENTIFIER: False,
+                      FEATURE_TOGGLE_MUTE: False}
+        self.support_select_source = False
+        self.support_volume_level = False
```
Just remove the support flag for: `volume`. I'll leave comments below how to change the current code.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,184 @@ def update_state(self, new_state):
+                                       CHAR_VOLUME_CONTROL_TYPE,
+                                       CHAR_VOLUME_SELECTOR))
+            if features & SUPPORT_VOLUME_SET:
+                self.support_volume_level = True
```
This could then be removed as well

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,184 @@ def update_state(self, new_state):
+            self.char_mute = serv_speaker.configure_char(
+                CHAR_MUTE, value=False, setter_callback=self.set_mute)
+
+            volume_control_type = 1 if self.support_volume_level else 2
```
```suggestion
            volume_control_type = 1 if CHAR_VOLUME in self.chars_speaker else 2
```
You might need to add a line break

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,184 @@ def update_state(self, new_state):
+
+        if self.support_select_source:
+            self.sources = self.hass.states.get(self.entity_id).attributes.get(
+                ATTR_INPUT_SOURCE_LIST)
```
Maybe put the line break after `self.entity_id)` instead?
```py
            self.sources = self.hass.states.get(self.entity_id) \
                .attributes.get(ATTR_INPUT_SOURCE_LIST)
```

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,184 @@ def update_state(self, new_state):
+                CHAR_ACTIVE_IDENTIFIER, setter_callback=self.set_input_source)
+            for index, source in enumerate(self.sources):
+                serv_input = self.add_preload_service(
+                    SERV_INPUT_SOURCE, [CHAR_IDENTIFIER, CHAR_NAME])
```
```suggestion
                    SERV_INPUT_SOURCE, (CHAR_IDENTIFIER, CHAR_NAME))
```
You should be able to use a tuple here.

adrum replied:
When I did this, my tests started failing.

cdce8p replied:
Missed that, ignore my initial comment

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,184 @@ def update_state(self, new_state):
+        """Send volume step value if call came from HomeKit."""
+        _LOGGER.debug('%s: Step volume by %s',
+                      self.entity_id, value)
+
```
Empty line

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,184 @@ def update_state(self, new_state):
+        """Send input set value if call came from HomeKit."""
+        _LOGGER.debug('%s: Set active_identifer to %s',
+                      self.entity_id, value)
+
```
Empty line

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,184 @@ def update_state(self, new_state):
+    def set_remote_key(self, value):
+        """Send remote key value if call came from HomeKit."""
+        _LOGGER.debug('%s: Set remote key to %s', self.entity_id, value)
+
```
Empty line

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,184 @@ def update_state(self, new_state):
+            CHAR_ACTIVE, setter_callback=self.set_on_off)
+
+        if CHAR_REMOTE_KEY in self.chars_tv:
+            self.char_remote_key = serv_tv.configure_char(
```
You don't need to create a local char var if it isn't updated from HA. `serv_tv.configure_char(...` is enough

adrum replied:
I was originally saving these off for the tests. I removed them for now, unless you think we will need them for the tests.

cdce8p replied:
You're right here again. I didn't though about the tests.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,184 @@ def update_state(self, new_state):
+            serv_speaker.configure_char(CHAR_VOLUME_CONTROL_TYPE,
+                                        value=volume_control_type)
+
+            self.char_volume_selector = serv_speaker.configure_char(
```
Same here: `serv_tv.configure_char(...`

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,184 @@ def update_state(self, new_state):
+                CHAR_VOLUME_SELECTOR, setter_callback=self.set_volume_step)
+
+            if CHAR_VOLUME in self.chars_speaker:
+                self.char_volume = serv_speaker.configure_char(
```
And here: `serv_tv.configure_char(...`

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,184 @@ def update_state(self, new_state):
+
+            if CHAR_VOLUME in self.chars_speaker:
+                self.char_volume = serv_speaker.configure_char(
+                    CHAR_VOLUME, setter_callback=self.set_volume)
```
You removed the callback method. See my first comment

cdce8p reviewed (changes requested):
We're getting close here. I haven't looked at the tests yet, do you want/need to updated them?

As a side note, do you run local checks?
That way you could remove the need for lint fixes after the fact. Since `tox` is pretty slow, take a look at testing outside tox: [doc](https://developers.home-assistant.io/docs/en/development_testing.html#testing-outside-of-tox). Especially I would recommend the following:
```bash
# (Activate virtual environment)
# Only use 'requirements_test', not 'test_all'
pip3 install -r requirements_test.txt
# Install HAP-python manually
pip3 install HAP-python

# When I test, I usually run:
script/lint
pytest tests/components/homekit/
```

adrum commented:
@cdce8p Thank you so much for sending that info over. I've been running tox locally and it is not ideal. I figured out how to specify targets and files, but it's still pretty slow. This will help greatly.

I will take a look at your review in a few hours. I will update the tests to accommodate the changes. I will get those updated after I implement the review changes.

cdce8p commented:
@adrum Sounds good to me. As for the local tests, I figured that you were using tox since I had the exact same issue once. It's just so slow; the new way is way faster though, usually sub 30s for both comments 😁

adrum committed:
fixes

adrum committed:
fixed tests

adrum committed:
removed volume flag

adrum committed:
missed one

cdce8p reviewed (changes requested):
It seems I show take a closer look at the suggestions I'm making :sweat_smile:
Good catches

adrum committed:
added code coverage

adrum commented:
No problem! I think I've got the tests update so it covers the whole file now. We shall see. I would say go ahead check out the tests whenever you feel the main file is ready.

adrum committed:
added volume level set test

adrum committed:
test lack of input sources

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,186 @@ def update_state(self, new_state):
+        super().__init__(*args, category=CATEGORY_TELEVISION)
+
+        self._flag = {CHAR_ACTIVE: False, CHAR_ACTIVE_IDENTIFIER: False,
+                      FEATURE_TOGGLE_MUTE: False}
```
I would use `CHAR_MUTE` as flag key
```suggestion
                      CHAR_MUTE: False}
```

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,186 @@ def update_state(self, new_state):
+        """Move switch state to value if call came from HomeKit."""
+        _LOGGER.debug('%s: Set switch state for "toggle_mute" to %s',
+                      self.entity_id, value)
+        self._flag[FEATURE_TOGGLE_MUTE] = True
```
```suggestion
        self._flag[CHAR_MUTE] = True
```

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,186 @@ def update_state(self, new_state):
+        # Power state television
+        hk_state = current_state not in (STATE_OFF, STATE_UNKNOWN)
+        if not self._flag[CHAR_ACTIVE]:
+            hk_state = 1 if hk_state else 0
```
This line is redundant. `Ture` is interpreted as `1` already.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,186 @@ def update_state(self, new_state):
+                _LOGGER.debug('%s: Set current mute state to %s',
+                              self.entity_id, current_mute_state)
+                self.char_mute.set_value(current_mute_state)
+            self._flag[FEATURE_TOGGLE_MUTE] = False
```
```suggestion
            self._flag[CHAR_MUTE] = False
```

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,186 @@ def update_state(self, new_state):
+        if CHAR_VOLUME_SELECTOR in self.chars_speaker:
+            current_mute_state = new_state.attributes.get(
+                ATTR_MEDIA_VOLUME_MUTED)
+            if not self._flag[FEATURE_TOGGLE_MUTE]:
```
```suggestion
            if not self._flag[CHAR_MUTE]:
```

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,186 @@ def update_state(self, new_state):
+                    index = self.sources.index(source_name)
+                    self.char_input_source.set_value(index)
+                else:
+                    self.char_input_source.set_value(0)
```
During normal use this shouldn't happen. However if it does, it will create bug reports.
Can you add a `Warning` above to say that the source list got out of sync and the user should restart HA?

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +171,186 @@ def update_state(self, new_state):
+        """Send remote key value if call came from HomeKit."""
+        _LOGGER.debug('%s: Set remote key to %s', self.entity_id, value)
+        if value in MEDIA_PLAYER_KEYS:
+            service = MEDIA_PLAYER_KEYS[value]
```
For the two lines above:
```py
service = MEDIA_PLAYER_KEYS.get(value)
if service:
    ...
```

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -1,23 +1,52 @@
 
 _LOGGER = logging.getLogger(__name__)
 
+REMOTE_KEYS = {
```
`REMOTE_KEYS` isn't used, is it?
If it's just for documenting proposes, add the keys to `MEDIA_PLAYER_KEYS` as comments:
```py
MEDIA_PLAYER_KEYS = {
  # 0: "Rewind",
  # 1: "FastForward",
  11: SERVICE_MEDIA_PLAYER_PAUSE,
  # ...
}

adrum replied:
My plan is to add a linked remote entity later on, which will use these keys. It will help me distinguish which service+entity needs called.

cdce8p replied:
Yeah, no problem. Just add those as comments inside `MEDIA_PLAYER_KEYS`. I would like to avoid having unused variables in the finished code.

cdce8p reviewed (changes requested):
Just some last small changes for the accessory. Will look at the tests next

cdce8p commented on tests/components/homekit/test_type_media_players.py:
```diff
@@ -129,3 +134,160 @@
+    assert acc.category == 31  # Television
+
+    assert acc.char_active.value == 0
+    assert acc.char_volume.value == 0
```
Integrate the test for `char_volume` within the first one.

adrum committed:
more changes

cdce8p commented on tests/components/homekit/test_type_media_players.py:
```diff
@@ -129,3 +134,160 @@
+    assert acc.char_mute.value is False
+
+    with raises(AttributeError, match="no attribute 'char_input_source'"):
+        acc.char_input_source
```
I wouldn't check this condition that way. Instead create a new test / modify this one to setup a basic media_player that only supports on/off and play/pause. Then assert that the `self.chars_` dictionaries are and `self.support_select_source` are correct.
Lastly I would check that no errors are raised when the state changes and `update_state` is run.

cdce8p commented on tests/components/homekit/test_type_media_players.py:
```diff
@@ -129,3 +134,160 @@
+    hass.states.async_set(entity_id, None, {ATTR_DEVICE_CLASS: DEVICE_CLASS_TV,
+                                            ATTR_SUPPORTED_FEATURES: 3468,
+                                            ATTR_MEDIA_VOLUME_MUTED: False,
+                                            })
```
```py
    hass.states.async_set(entity_id, None, {
        ATTR_DEVICE_CLASS: DEVICE_CLASS_TV, ATTR_SUPPORTED_FEATURES: 3468,
        ATTR_MEDIA_VOLUME_MUTED: False})
```

cdce8p reviewed (changes requested):
Some comments regarding the tests.

cdce8p commented on homeassistant/components/homekit/type_media_players.py:
```diff
@@ -142,3 +168,187 @@ def update_state(self, new_state):
+    def set_remote_key(self, value):
+        """Send remote key value if call came from HomeKit."""
+        _LOGGER.debug('%s: Set remote key to %s', self.entity_id, value)
+        service = MEDIA_PLAYER_KEYS[value]
```
```suggestion
        service = MEDIA_PLAYER_KEYS.get(value)
```

adrum committed:
added more tests

cdce8p committed:
Last changes

cdce8p commented:
I've just pushed some last changes, but it seems as if it's finally ready. What do you think?

adrum commented:
@cdce8p This looks good to me! I was unaware of caplog. I will definitely keep that noted for future tests.

cdce8p commented:
@adrum Thanks for the great work you did to get this feature ready :1st_place_medal: :dolphin:
